### PR TITLE
[WIP] MAT Layers: Attention Modules

### PR DESCRIPTION
**Pull request by atreyamaj**
# Pull Request Template
These are half the layers required for the Molecular Attention Transformer.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by atreyamaj** [comment]
The tests (test_layers.py) I've written are currently not working as intended, thus for now this is a WIP and I've committed the layers. I'll push the other half soon as well while working on the tests.

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -47,3 +76,427 @@ def __init__(self, scale, eps=1e-5):
...
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> attention = dc.models.torch_models.layers.MATAttention('softmax', 0.33, 0.33)
+  >>> self_attn_layer = dc.models.torch_models.layers.MultiHeadedAttention(8, 1024, 0.1, attention)
+  >>> feed_fwd_layer = dc.models.torch_models.layers.PositionwiseFeedForward(d_input = 1024, activation = torch.nn.ReLU(), n_layers = 1, dropout = 0.1)
+  >>> block = dc.models.torch_models.layers.Encoder(self_attn_layer = self_attn_layer, feed_fwd_layer = feed_fwd_layer, d_model = 1024, dropout = 0.0, N = 3)
+  """
+
+  def __init__(self, self_attn_layer, feed_fwd_layer, d_model, dropout, N):
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:100]: Would not recommend passing layers as input arguments to the encoder block. It would be better to instead construct the `self_attn_layer` and `feed_fwd_layer` inside the `MATEncoder`. 
I would strongly advocate for renaming `d_model` and `dropout` to something found more consistently within DeepChem (for example `hsize` and `dropout_p`.

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -47,3 +76,427 @@ def __init__(self, scale, eps=1e-5):
   def forward(self, x):
     norm = self.scale / torch.norm(x, dim=-1, keepdim=True).clamp(min=self.eps)
     return x * norm
+
+
+class Encoder(nn.Module):
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:81]: Rename to `MATEncoder` and change documentation to reflect that too.

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -47,3 +76,427 @@ def __init__(self, scale, eps=1e-5):
...
+      Number of identical layers to be stacked.
+    """
+
+    super(Encoder, self).__init__()
+    layer = EncoderLayer(
+        self_attn_layer=self_attn_layer,
+        feed_fwd_layer=feed_fwd_layer,
+        d_model=d_model,
+        dropout=dropout)
+    self.layers = clones(layer, N)
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:123]: Not sure we need the `clones` function here. You can just do
```
self.layers = nn.ModuleList([EncoderLayer(*args) for _ in range(N)])
```

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -47,3 +76,427 @@ def __init__(self, scale, eps=1e-5):
...
+    ----------
+    x: torch.Tensor
+      Input tensor.
+    sublayer: nn.Module
+      Layer whose output for normalized x will be added to x.
+    """
+    return x + self.dropout(sublayer(self.norm(x)))
+
+
+class EncoderLayer(nn.Module):
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:190]: Rename to `MATEncoderLayer`

**Comment by vsomnath** [comment]
This PR is going to need a bit of work, I will review in greater detail in a while.

File: deepchem/models/torch_models/layers.py (hunk 1 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:100]: I have made the changes for `hsize` and `dropout_p`.

For the input arguments to the block, I can maybe remove everything and only keep `N` and one input layer: `MATEncoderLayer`?

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:81]: Done!

File: deepchem/models/torch_models/layers.py (hunk 4 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:190]: Done!

File: deepchem/models/torch_models/layers.py (hunk 3 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py]: Done!

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -14,19 +16,19 @@ class ScaleNorm(nn.Module):
...
+
   Examples
   --------
   >>> import deepchem as dc
   >>> scale = 0.35
-  >>> layer = dc.models.torch_models.ScaleNorm(scale)
-  >>> input_tensor = torch.tensor([[1.269, 39.36], [0.00918, -9.12]])
-  >>> output_tensor = layer.forward(input_tensor)
+  >>> layer = dc.models.torch_models.layers.ScaleNorm(scale)
+  >>> input_tensor = torch.Tensor([[1.269, 39.36], [0.00918, -9.12]])
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:30]: `torch.tensor` is the function making the tensor, `torch.Tensor` is the type

File: deepchem/models/torch_models/layers.py (hunk 1 above)
Comment (vsomnath) [reply deepchem/models/torch_models/layers.py]: Do you have a good reason for why you pass a layer object directly as input to the model?

File: deepchem/models/torch_models/layers.py (hunk 5 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py]: Done!
